Use correct n_components in case n_components is passed as a float between 0 and 1

krishnachaitanya7 opened the pull request:
`n_components` is the total number of principal components. When n_components is passed as a float between 0 and 1, n_components is converted into integer something like this:
`n_components = np.searchsorted(ratio_cumsum, n_components) + 1`
There is a catch in numpy's searchsorted function. It returns index i where `a[i-1] < v <= a[i]`. So here `i` can be returned such that `v < a[i]` which is exactly what is mentioned in the documentation: 

If `0 < n_components < 1` and `svd_solver == 'full'`, select the number of components such that the amount of variance that needs to be explained is greater than the percentage specified by n_components. [[Link](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html)]

Here v is percentage specified by n_components, which comes as input from the user something like `PCA(n_components=0.99, svd_solver='full')` and a[i] is the amount of the variance that needs to be explained. So adding 1 here is redundant as v is ALREADY less than a[i] which is exactly what's required according to the above-mentioned documentation.

But if `v = a[i]` then 1 should be added as v is NOT less than a[i].

Hence I feel using np.searchsorted for this task isn't correct and it should be replaced by alternatives like:
`n_components = np.nonzero(ratio_cumsum > n_components)[0][0]`

Now, n_components is just the exact total number of principal components required.

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
It fixes the variable `n_components` which was calculated to be a bit higher before due to the addition of 1. This pull request aims to address that issue by changing that one line of code responsible for it. After this change, the number of principal components will be exactly how much is required.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

krishnachaitanya7 committed:
Use correct n_components

Before n_components was updated something like this:
n_components = np.searchsorted(ratio_cumsum, n_components) + 1
There is a catch in numpy's searchsorted function. It returns index i where a[i-1] < v <= a[i]. So here i can be returned such that v < a[i] which is exactly what is mentioned in the documentation: 
If 0 < n_components < 1 and svd_solver == 'full', select the number of components such that the amount of variance that needs to be explained is greater than the percentage specified by n_components.
so adding 1 here is redundant. 

But if v = a[i] then 1 should be added.

Hence I feel using np.searchsorted for this task isn't correct and it should be replaced by alternatives like:
n_components = np.nonzero(ratio_cumsum > n_components)[0][0]

Now, n_components is just the exact total number of principal components required.

jnothman commented on sklearn/decomposition/_pca.py:
```diff
@@ -464,7 +464,7 @@ def _fit_full(self, X, n_components):
             # variance percentage is superior to the desired threshold
             ratio_cumsum = stable_cumsum(explained_variance_ratio_)
-            n_components = np.searchsorted(ratio_cumsum, n_components) + 1
+            n_components = np.nonzero(ratio_cumsum > n_components)[0][0]
```
perhaps `np.argmax(ratio_cumsum > n_components)`?

but maybe there's a neater solution still

krishnachaitanya9 replied:
Hi @jnothman! Yes, your solution should work too. I have checked on my end it works perfectly. Also, your solution is much neater! Thank you so much for suggesting that.

krishnachaitanya9 replied:
And I also think it's correct as many tests would be failing, because now lesser n_components are calculated, which messes up output matrix shape, hence many tests are bound to fail

jnothman reviewed (commented):
This is causing test failures at present...

krishnachaitanya7 committed:
Edit my previous commit according to @jnothman's comment

Convert `n_components = np.nonzero(ratio_cumsum > n_components)[0][0]` to `n_components = np.argmax(ratio_cumsum > n_components)`

glemaitre commented:
I don't think that your analysis is correct. `n_component` will be the index + 1.

```python
>>> import numpy as np
>>> from sklearn.datasets import load_iris
>>> from sklearn.decomposition import PCA                                       

>>> X, y = load_iris(return_X_y=True)                                                                                    
>>> pca = PCA().fit(X)                                                           

>>> print(pca.explained_variance_ratio_)                                                
array([0.92461872, 0.05306648, 0.01710261, 0.00521218])

>>> print(pca.explained_variance_ratio_.cumsum())                                      
array([0.92461872, 0.97768521, 0.99478782, 1.        ])

>>> print(np.searchsorted(pca.explained_variance_ratio_.cumsum(), 0.95))                
1

>>> pca = PCA(n_components=0.95).fit(X)                                         
>>> print(pca.n_components_)                                                           
2
```

In the above snippet, we want to find the number of components such that the cumulative explained variance is greater than 0.95. The index found is `1` meaning that it requires 2 components.

krishnachaitanya9 commented:
Hi @glemaitre, your answer is spot on and I found my mistake. Thank you so much. Now another doubt in the implementation looms me. Please consider the below snippet:
```python
>>> import numpy as np
>>> from sklearn.datasets import load_iris
>>> from sklearn.decomposition import PCA 
>>> X, y = load_iris(return_X_y=True)

>>> print(X.shape)
(150, 4)
# Above line shows that there are 4 features in the dataset

>>> pca = PCA().fit(X)   
>>> print(pca.explained_variance_ratio_.cumsum())
[0.92461872 0.97768521 0.99478782 1.        ]
# Above line prints the cumulative sum

>>> pca = PCA(n_components=0.99478782);X_new = pca.fit_transform(X);print(X_new.shape)
(150, 4)
>>> print(pca.n_components_)
4
# Above line prints 4. Which is correct according to documentation.
# Amount of variance explained = 1.0
# n_components % = 0.99478782
# 1.0 > 0.99478782

>>> pca = PCA(n_components=0.97768521);X_new = pca.fit_transform(X);print(X_new.shape)
(150, 3)
>>> print(pca.n_components_)
3
# Above line prints 3. Which is correct according to documentation.
# Amount of variance explained = 0.99478782
# n_components % = 0.97768521
# 0.99478782 > 0.97768521

>>> pca = PCA(n_components=0.92461872);X_new = pca.fit_transform(X);print(X_new.shape)
(150, 1)
>>> print(pca.n_components_)
1
# Above line prints 1. Which I think is not consistent with documentation.
# Amount of variance explained = 0.92461872
# n_components % = 0.92461872
# 0.92461872 = 0.92461872 (which should be greater, not equal to)
# The actual n_components should have been 2.
```
Now no one would put such exact n_component % in real life. Please correct me if I am wrong. Thank you so much for your time.
